### PR TITLE
[DISCUSSION] feat: allow server-side modification of user doc on account creation

### PR DESCRIPTION
**Pull request by janl**
Heya, I’m using this outside of Hoodie and I found the need to be able to change the user doc on account creation.

E.g. I need to add another role to the user doc, and the server will know which one.

My idea was to add a function, for illustration purposes called `userDocModificationFun` (please bikeshed), to pass into the plugin options, and then, if that function exists, it gets called on `account.add()`. In order to support this, the top level options object needs to be passed down to `account.add()` (which looks like was planned anyway, this adds the plumbing).

Please let me know if this is an acceptable addition. If yes, I’m happy to add tests and clean things up.

### CODE REVIEW

**Comment by gr2m** [comment]
When you say you use `hoodie-account-server` outside Hoodie, can you explain why you can’t modify what you pass to `account.add()` directly?

**Comment by janl** [comment]
@gr2m I’m using the hoodie-account-client as well.

**Comment by janl** [comment]
I could pass these things in from before hoodie-account-client, but I can’t trust the client about these things.

**Comment by gr2m** [comment]
Ah, what you need is a hook like `pre:add` where you can alter the account object before it gets stored to the users database? We have something similar for the client API: https://github.com/hoodiehq/hoodie-client/blob/issue/3/lib/init.js#L6-L10 It's currently undocumented because we are not 100% happy with the API, we could do something similar for the server side, too. 

One idea would be to see how hapi is doing it with their request lifecycle. We also had a discussion about a hook dream API here: https://github.com/hoodiehq/hoodie-client/issues/42, which might work for your use case as well. 

tl;dr;

- the topic is relevant
- we don’t have a simple solution
- we can add an undocumented API for the time being so you can move forward

**Comment by gr2m** [comment]
just an idea for a workaround, you could maybe use https://github.com/pouchdb/pouchdb-wrappers to wrap the PouchDB instance that you pass to hoodie-account-server?

**Comment by janl** [comment]
I like the idea of hooks and other things, but I’ve now realised, that the information I’m storing is only available at request-time (e.g. `request.headers.host`), so I need to figure out how to be able to access that, or pass that down. Will reopen, when I’ve got something.

**Comment by janl** [comment]
well that’s embarrassing 😳 here’s a version that does that :D, so my stuff can be made to work (squashed and force pushed)

File: api/accounts/add.js [hunk 1]
```diff
@@ -25,6 +25,13 @@ function addAccount (state, properties, options) {
     ].concat(properties.roles || [])
   }
 
+  var hooks = options.hooks || {}
+
+  if (hooks.account && hooks.account.beforeAdd &&
+    typeof hooks.account.beforeAdd === 'function') {
+    doc = hooks.account.beforeAdd(doc, request)
```
Comment (janl) [inline api/accounts/add.js]: we’d make this a `util` of course

File: routes/account.js [hunk 2]
```diff
@@ -41,7 +41,7 @@ function accountRoutes (server, options, next) {
         password: password,
         include: query.include,
         id: id
-      })
+      }, options, request)
```
Comment (gr2m) [inline routes/account.js]: the [accounts API](https://github.com/hoodiehq/hoodie-account-server/tree/master/api) is somewhat independent from routes, it could even be moved to an independent module. Ideally it will match the [admin client API for accounts](https://github.com/hoodiehq/hoodie-account-client/tree/master/admin).

I would not pass in route options and the request into the `accounts.add` method, instead I’d run `hooks.account.beforeAdd` here, before running `accounts.add`. Would that work?

If I recall correctly, the database in which the user account document shall be stored, depends on the request, too? Or is that not the case any longer?

**Comment by janl** [comment]
The doc I want to augment is created in [`api/account/add.js`](https://github.com/hoodiehq/hoodie-account-server/blob/master/api/accounts/add.js#L18-L26), so running the hook before that doesn’t really work. Although I like the idea of such a design. Maybe we can pass a prototype doc into `addAccount()` that gets merged with the default defined in there?

**Comment by gr2m** [comment]
We had a discussion on this and will see if we can use the custom "signup" route handler for this. This feature is not yet implemented. The idea is that besides custom requests that can be passed as on object to the [`@hoodie/acocunt-server` hapi plugin](https://github.com/hoodiehq/hoodie-account-server/tree/master/plugin#optionsrequests), you can also pass in account routes for `signup`, `signin`, `signout` and `destroy`, which would then overwrite the default route handlers from [routes/account.js](https://github.com/hoodiehq/hoodie-account-server/blob/master/routes/account.js). It will mean some code duplication, but I see this as a power user tool, and the benefit would be minimal added complexity to `@hoodie/acocunt-server` as it is today

**Comment by gr2m** [comment]
@janl we now use [before-after-hook](https://github.com/gr2m/before-after-hook) at several places in Hoodie. I think that could fit perfectly with your initial requirement

> Heya, I’m using this outside of Hoodie and I found the need to be able to change the user doc on account creation.
> 
> E.g. I need to add another role to the user doc, and the server will know which one.

We could for example expose an `add` hook, which would wrap this [`db.put(doc)` call](https://github.com/hoodiehq/hoodie-account-server-api/blob/943b55b1ac7cacd12977a00245d82c018d01d100/lib/accounts/add.js#L36) so that you could do something like this

```js
server.plugins.account.api.hook.before('add', function (properties) {
  properties.roles.push('mycustomrole')
})
```

`server.plugins.account.api` is exposed by the `@hoodie/account-server` hapi plugin, so you can call it after registration is done. We could also add a `hooks` option to the plugin itself, it could work like this:

```js
server.register({
  register: hapiAccount,
  options: {
    // other options
    hooks: {
      add: {
        before: function (properties) { properties.roles.push('mycustomrole') }
      }
    }
  }
}, function (error) {
  if (error) {
    throw error
  }

  // plugin account api, see below
});
```

Would that be helpful for your use case?
